### PR TITLE
RUL-47: add concatenate action rule

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description**

Add and describe the new *concatenate* action in rule engine

**Definition Of Done**

| Q                                 | A
| --------------------------------- | ---
| Technical Review and 2 GTM        | Todo
| English Review and 1 GTM          | Todo

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 1]
```diff
@@ -87,7 +87,7 @@ An enrichment rule is structured as follow:
             - type:*
               [Diverse elements according to the action]
 
-Elements with * are mandatory. Fill in the locale and scope elements only if your condition applies on localizable and/or scopable attributes. 
+Elements with * are mandatory. Fill in the locale and scope elements only if your condition applies on localizable and/or scopable attributes.
 
 **Dashes** (-) must be placed before an element field and after each element contained in value part.
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
**Dashes** (-) must be placed before an element field and after each element contained in the value part.
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 2]
```diff
@@ -195,7 +195,7 @@ Enrichment Rule Definition
 Available Actions List
 ++++++++++++++++++++++
 
-Akeneo rules engine proposes 4 kinds of actions:
+Akeneo rules engine proposes 5 kinds of actions:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
Akeneo rules engine enables 5 kinds of actions:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 3]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
                   field: categories
                   value: []
 
+Concatenate
+___________
+
+This action concatenates at least two values into a single value. A space separates each source values.
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
This action concatenates at least two values into a single value. A space separates each source value.
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 4]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+ - metric
+ - number
+ - price collection
+ - simple or multi select (option codes)
+
+The possible target attribute types are:
+ - text
+ - text area
+
+**The parameters from and to are required in the format. Depending of source attribute type, some optional keys can be additionally set:**
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
**The parameters from and to are required in the format. Depending on the source attribute type, some optional keys can be set:**
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 5]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+
++------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| from | List of sets for all attribute types:                                                                                                                                      |
+|      |                                                                                                                                                                            |
+|      | - field: attribute code.                                                                                                                                                   |
+|      | - locale: locale code for which value is assigned (optional).                                                                                                              |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | - locale: locale code for which the value is assigned (optional).                                                                                                              |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 6]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+|      | - scope: channel code for which value is assigned (optional).                                                                                                              |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | - scope: channel code for which the value is assigned (optional).                                                                                                              |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 7]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+|      |                                                                                                                                                                            |
+|      | For date attribute:                                                                                                                                                        |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | For date attributes:                                                                                                                                                        |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 8]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+|      |                                                                                                                                                                            |
+|      | - format: format of the date following the `PHP format specification <https://www.php.net/manual/en/function.date.php>`_. Optional. Default is *Y-m-d* (e.g. *2020-01-31*) |
+|      |                                                                                                                                                                            |
+|      | For price collection attribute:                                                                                                                                            |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | For price collection attributes:                                                                                                                                            |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 9]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+|      |                                                                                                                                                                            |
+|      | - currency: currency code for which price is assigned. Optional. By default all the prices in the collection are displayed, separated by a coma.                           |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | - currency: currency code for which the price is assigned. Optional. By default all the prices in the collection are displayed, separated by a coma.                           |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 10]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
++------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| to   | One set of:                                                                                                                                                                |
+|      |                                                                                                                                                                            |
+|      | - field: attribute code.                                                                                                                                                   |
+|      | - locale: locale code for which value is assigned (optional).                                                                                                              |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | - locale: locale code for which the value is assigned (optional).                                                                                                              |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 11]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+|      | - scope: channel code for which value is assigned (optional).                                                                                                              |
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
|      | - scope: channel code for which value is assigned (optional).                                                                                                              |
```
```suggestion
|      | - scope: channel code for which the value is assigned (optional).                                                                                                              |
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 12]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
++------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+
+.. tip::
+
+    For instance, to concatenate the brand (non localizable and non scopable) and the model in en_US locale into the description value in en_US locale, action will be as follows:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
    For instance, to concatenate the brand (non-localizable and non-scopable) and the model in en_US locale into the description value in en_US locale, the action will be as follows:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 13]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+            - type: concatenate
+              from:
+                - field: brand
+                - field: model
+                  locale: en_US
+              to:
+                field: description
+                locale: en_US
+
+    To concatenate the model in en_US locale, the color in en_US locale and the year of the release date into the title value in en_US locale, action will be as follows:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
    To concatenate the model in en_US locale, the color in en_US locale and the year of the release date into the title value in en_US locale, the action will be as follows:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 14]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+                - field: model
+                  locale: en_US
+                - field: color
+                - field: release_date
+                  format: Y
+              to:
+                field: title
+                locale: en_US
+
+    To concatenate the model in en_US locale and the price in USD and in mobile channel into the subtitle value in en_US locale and mobile channel, action will be as follows:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]:
```suggestion
    To concatenate the model in en_US locale and the price in USD and in the mobile channel into the subtitle value in en_US locale and mobile channel, the action will be as follows:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 15]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+ - text
+ - text area
+
+**The parameters from and to are required in the format. Depending on the source attribute type, some optional keys can be set:**
+
++------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| from | List of sets for all attribute types:                                                                                                                                      |
+|      |                                                                                                                                                                            |
+|      | - field: attribute code.                                                                                                                                                   |
+|      | - locale: locale code for which the value is assigned (optional).                                                                                                          |
```
Comment (pierallard) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]: I may add "for localizable attributes"
(same for other ones)

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 16]
```diff
@@ -336,6 +336,98 @@ Two parameters are required while the three others are optional.
...
+ - text
+ - text area
+ - date
+ - identifier
+ - metric
+ - number
+ - price collection
+ - simple or multi select (option codes)
+
+The possible target attribute types are:
```
Comment (nicomaligne) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst]: target attributes ?

Comment (nmarniesse) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst]: nope it's the word "types" that must be pluralize not the attribute (it's just an adjective in this case)

**Pull request merged**